libs/full/parcelport_lci/include/hpx/parcelport_lci/sender_connection.hpp

```diff
@@ -152,85 +170,39 @@ namespace hpx::parcelset::policies::lci {
                 }
                 HPX_ASSERT(long_msg_num == i);
             }
-            else
-            {
-                sync_ = nullptr;
-            }
-
-            state_ = initialized;
-
-            handler_ = HPX_FORWARD(Handler, handler);
-
-            if (!send())
-            {
-                postprocess_handler_ =
-                    HPX_FORWARD(ParcelPostprocess, parcel_postprocess);
-                add_connection(sender_, shared_from_this());
-            }
-            else
-            {
-                HPX_ASSERT(!handler_);
-                error_code ec;
-                parcel_postprocess(ec, there_, shared_from_this());
-            }
+            return header_;
         }
 
         bool send()
-        {
-            switch (state_)
-            {
-            case initialized:
-                return send_header();
-
-            case sent_header:
-                return done();
-
-            default:
-                HPX_ASSERT(false);
-            }
-
-            return false;
-        }
-
-        bool send_header()
         {
             if (iovec.count == 0)
             {
-                if (LCI_putma(util::lci_environment::h_endpoint(),
-                        iovec.piggy_back, dst_rank, 0,
-                        LCI_DEFAULT_COMP_REMOTE) != LCI_OK)
+                while (LCI_putma(util::lci_environment::get_endpoint(),
+                           iovec.piggy_back, dst_rank, 0,
+                           LCI_DEFAULT_COMP_REMOTE) != LCI_OK)
                 {
-                    return false;
+                    continue;
                 }
+                return true;
             }
             else
             {
-                if (LCI_putva(util::lci_environment::h_endpoint(), iovec, sync_,
-                        dst_rank, 0, LCI_DEFAULT_COMP_REMOTE,
-                        nullptr) != LCI_OK)
+                auto* sharedPtr_p =
```

Comment:
I've never seen anybody creating a pointer to a `std::shared_ptr` before. Is that an oversight? Wouldn't using a plain `std::shared_ptr` sufficient?

Reply:
It is intentional. The C library is not aware of a shared_ptr. We can only pass a plain void* to LCI_putva as a "user-defined operation context". This pointer will be returned to the user via the completion queue after the LCI_putva completes. If we don't have a shared_ptr in the heap to keep a reference to this object, it will be deallocated right after this async_write function return (because no other shared_ptr is pointing to it).  Basically, I need a way to ensure the sender_connection object is not deallocated after we post LCI_putva and before we get it back from the completion queue. Do you have a better way than this?

Reply:
Understood. I have no better solution at this point.